common/src/PCLPointCloud2.cpp

```diff
@@ -0,0 +1,235 @@
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id$
+ *
+ */
+
+#include <vector>
+
+#include <pcl/common/io.h>
+#include <pcl/pcl_macros.h>
+#include <pcl/exceptions.h>
+#include <pcl/PCLPointCloud2.h>
+
+bool
+pcl::PCLPointCloud2::concatenate (pcl::PCLPointCloud2 &cloud1, const pcl::PCLPointCloud2 &cloud2)
+{
+  if (cloud1.is_bigendian != cloud2.is_bigendian)
+  {
+    // In future, it might be possible to convert based on pcl::getFieldSize(fields.datatype)
+    PCL_ERROR ("[pcl::PCLPointCloud2::concatenate] Endianness of clouds does not match\n");
+    return (false);
+  }
+
+  const auto size1 = cloud1.width * cloud1.height;
+  const auto size2 = cloud2.width * cloud2.height;
+  //if one input cloud has no points, but the other input does, just select the cloud with points
+  switch ((bool (size1) << 1) + bool (size2))
+  {
+    case 1:
+      cloud1 = cloud2;
+      PCL_FALLTHROUGH;
+    case 0:
+      PCL_FALLTHROUGH;
+    case 2:
+      cloud1.header.stamp = std::max (cloud1.header.stamp, cloud2.header.stamp);
+      return (true);
+    default:
+      break;
+  }
+
+  bool strip = true;
+  if (cloud1.fields.size () == cloud2.fields.size ())
+  {
+    // potentially we can do a simple memcpy
+    if (std::equal(cloud1.fields.begin (),
+                    cloud1.fields.end (),
+                    cloud2.fields.begin (),
+                    [](const auto& f1, const auto&f2)
+                    {
+                      return f1.name == f2.name;
+                    }))
+    {
+      strip = false;
+    }
+  }
+
+  struct FieldDetails
+  {
+    std::size_t idx1, idx2;
+    std::uint16_t size;
+    FieldDetails (std::size_t idx1_, std::size_t idx2_, std::uint16_t size_): idx1 (idx1_), idx2 (idx2_), size (size_)
+    {}
+  };
+  std::vector<FieldDetails> valid_fields;
+  const auto max_field_size = std::max (cloud1.fields.size (), cloud2.fields.size ());
+  valid_fields.reserve (max_field_size);
+
+  // refactor to return std::optional<std::vector<FieldDetails>>
+  if (strip)
+  {
+    std::size_t i = 0, j = 0;
+    while (i < cloud1.fields.size () && j < cloud2.fields.size ())
+    {
+      if (cloud1.fields[i].name == "_")
+      {
+        ++i;
+        continue;
+      }
+      if (cloud2.fields[j].name == "_")
+      {
+        ++j;
+        continue;
+      }
+
+      // We're fine with the special RGB vs RGBA use case
+      if ((cloud1.fields[i].name == "rgb" && cloud2.fields[j].name == "rgba") ||
```

Comment:
I propose to extract this check into a lambda variable and reuse in line 82.

Reply:
Oh yes. For sure